### ISSUE TITLE
Reliably Fetch Account State

### ISSUE THREAD

**Issue by bh2smith** [gnosis/dex-services#684]
**TL;DR** Smart Contract revert with "Amount exceeds user's balance.". A competing solver was attempting to spend pending balances from an already accepted solution to the current batch. When the solution was submitted and existing trades reverted the balances were no longer sufficient.

Last week our driver software experienced an interesting and unexpected failure to submit solution to Batch 5286263. The smart contract failed with the message "Amount exceeds user's balance." Note that, because the driver queries the return value of `submitSolution` before actually submitting the transaction, this error occurred without an existing failed transaction to be linked here.

After careful investigation, it was determined that a competing solver (our fallback-solver) was working off of an incorrect state of accounts. In particular, there has already been a solution submitted before the competing solver had fetched or queried the `Orderbook` (i.e. account balances and active orders). The accounts that the solver was working from, were those that had already bee updated by an accepted solution. After finding another solution and submitting its proposal, the smart contract internally reverted the previously accepted solution (hence, reverting the balances to their previous state) and suddenly, the solution proposed by the competing solver was no longer valid (due to insufficient balance). For more details on the issue with links to relevant transactions and solutions search for "Batch 5286263" in our [oncall report](https://docs.google.com/document/d/1WjlqerD4P1Tn-oLmaCZiPWFZDtiyh25ElVTHmwzNgP8/edit#)

There are likely a few possible way to tackle this new problem, so I will propose here a few thoughts and see how the discussion develops.

One universal fact that must be considered in our solution is that competing solvers should remain independent. That is, since this fix will be made to the driver (i.e. from within `dex-services`) any independent application running an instance of this project will not know anything else about other instances.
  
**Some ideas**

1. Configure driver to fetch account state separately from orders so it is returned earlier. This may already be in place.
2. Configure driver to fetch orderbook from the first published block after batch has closed. 

_Note_, however, that it is technically possible that this first block published after batch has closed could already contain a successful solution submission. In fact, it is not even unlikely that this could be the case. For example, suppose there were two obviously matching orders placed at the beginning of the batch. It is now obvious that a non-trivial solution to this auction exists and could be submitted (and accepted) as early as the first published block after batch closure.

3. Configure driver to fetch account state from last block published before batch closure. 
4. Query for orderbook from block number minus one of an already accepted solution... seems like a game of cat and mouse.

**Comment by bh2smith** [comment]
I think that, if possible. Option 3 in combination with 1 would be a reasonable approach.

**Comment by fleupold** [comment]
I think an alternative solution would be to query the public `lastSolution` field from dex-services (or a wrapper contract) and if its batchId is the one we are currently trying to solve deduct its trades from the fetched balances.

This will introduce more logic in the driver, but seems like a robust and logically correct modification.

**Comment by fleupold** [comment]
We discussed that a good strategy to assure consistency would be twofold

1) Fetch `latestSolution` from the contract and (if it affects the current batch) adjust the sellTokenBalance returned
2) Subscribe to events before starting the orderbook fetch and if an event comes on restart the fetch. This bears the risk of running into an infinite loop (specifically as the time to fetch order becomes larger)

The problem about 1) is that the ABI only exposes the non array fields on latestSolution (batchId, solutionSubmitter, feeReward, objectiveValue) and thus we have no access to the solution from within the services.
We could build this logic into the BatchExchangeViewer, but this might significantly increase our gas usage in case we already have a competing solution.

**Comment by bh2smith** [comment]
Is there some way we could combine option 1 with knowledge of block number that solution was included so that we would know to query the account state from the block prior to the solution? Of course, it is also possible that two solutions had already been accepted. So this could get messy. 

I still feel like there must be some way to deterministically get the correct state of accounts based on system time and block numbers.

**Comment by fleupold** [comment]
Unfortunately, with the current smart contracts, there is no guaranteed way to get reliable balance from the smart contract directly (without adding extra logic to the Viewer). 

You can theoretically submit a solution in the first block (*b*) that accepts solutions for a given batch and thus there is not a single block you can query that returns the correct balances (block *b-1* returns the balances as of the previous batch, block *b* returns the balances with trades applied).

**Comment by josojo** [comment]
What is the current state of the issue?

It keeps on popping up more and more with the open-solver, as this solver starts quite late to fetch its orders.

PS: I am fine with just switching the whole infrastructure to this  order book reading via events and keep this issue unaddressed (it would only continue to be an issue in the back-up scenario when the event-based orderbook  fetching does not work)

**Comment by fleupold** [comment]
We are still working on the event based orderbook and given its complexity I'd prefer to have it run shadow traffic for a week. So this is likely going to take at least 2 more weeks.

The other option requires changes to the BatchExchangeViewer which have not yet been started.

**Comment by fleupold** [comment]
Actually it looks like we cannot get the trade information from latestSolution field on the smart contract. Here is the solidity type definition

```solidity
struct SolutionData {
    uint32 batchId;
    TradeData[] trades;
    uint16[] tokenIdsForPrice;
    address solutionSubmitter;
    uint256 feeReward;
    uint256 objectiveValue;
}
```

And here is the auto-generated method for the field has the following ABI

```json
{
  "constant": true,
  "inputs": [],
  "name": "latestSolution",
  "outputs": [
    {
      "name": "batchId",
      "type": "uint32"
    },
    {
      "name": "solutionSubmitter",
      "type": "address"
    },
    {
      "name": "feeReward",
      "type": "uint256"
    },
    {
      "name": "objectiveValue",
      "type": "uint256"
    }
  ],
  "payable": false,
  "stateMutability": "view",
  "type": "function"
},
```

It shows that only the static fields, but not the dynamic arrays for `trades` and `tokenIdsForPrice` are accessible on the contract's interface. This means that we cannot account for currently pending trades from inside the BatchExchangeViewer as we don't have access to their order information and amounts.
Even from our off-chain services doing so would require reading specific smart contract storage locations which is very slow and cumbersome.

cc-ing @nlordell  and @e00E to see if they have an idea of a workaround, but for if not I think we need to wait until the event based orderbook is ready.

**Comment by nlordell** [comment]
The only thing I can think of is to walk down the blocks of the block chain until it is either in the previous block or a `SolutionSubmission` event is found and then get the required information from the events in that block.

**Comment by e00E** [comment]
> Configure driver to fetch orderbook from the first published block after batch has closed.

Something like this can work. Instead of *starting* the batch late which causes the orderbook problem we can start immediately, fetch the orderbook, then wait however long we want to delay to submit the solution.

**Comment by nlordell** [comment]
> Something like this can work. Instead of starting the batch late which causes the orderbook problem we can start immediately, fetch the orderbook, 

We still run into issues with this since we have two schedulers, the system time one and the EVM one and it is possible that the EVM scheduler starts after the system time scheduler has already retrieved the orderbook and submitted a solution.

The issue with retrieving the orderbook early in all cases is that we run into the issue where we can't be certain that a new order won't come in at the last second.

I think this is a good solution to reduce the number of alerts in the channel, however, so it might be worth implementing anyway. This way the scheduler would be responsible for querying the orderbook.

**Comment by fleupold** [comment]
I also like this suggestion. Further down the line it might make sense for the scheduler to decide which jobs to restart so having the (in the future stream of) orders available could make sense.

At least for now it does however put a quite long running blocking task on the main thread.

**Comment by nlordell** [comment]
PR was mistakenly closed

**Comment by bh2smith** [comment]
@fleupold - was this closed because it is solved?

**Comment by fleupold** [comment]
https://github.com/gnosis/dex-services/pull/736 fixed this issue, by implementing something similar to your initial idea 2. We now fetch the orderbook on the last block that it was still collecting orders. At this point it will already contain all relevant order/deposits/withdraws...

**Comment by bh2smith** [comment]
That is excellent, although the way you describe it, it sounds more like the initial idea #3

> 3. Configure driver to fetch account state from last block published before batch closure.
